### ISSUE TITLE
Sensor array cannot compile anymore

### ISSUE THREAD

**Issue by rkierzek** [RobTillaart/MAX6675#11]
Some time ago, ive successfully used this library to read multiple sensors MAX6675. 
**Library version 0.1.2, 0.1.3 compiling and working fine.**
Library version 0.2.0, 0.3.0 i have compilation errors. First error is in declaration line.
I use array, and use in for() statements.
```cpp
// declaration
MAX6675 ThermoCouples[ThermoCouplesNum];

// initialisation
ThermoCouples[i].begin(MAXclockPin, THCselectPins[i], MAXmisoPin);
ThermoCouples[i].setSPIspeed(MAX6675SPIClockSpeed);

//access
int status = ThermoCouples[THCnumber].read();
int temperature = ThermoCouples[THCnumber].getTemperature()

```
(...)\MAX6675_softSPI_array_Serial_LCD\MAX6675_softSPI_array_Serial_LCD.ino:34:39: error: no matching function for call to 'MAX6675::MAX6675()'
 MAX6675 ThermoCouples[ThermoCouplesNum];
                                       ^
In file included from (...)MAX6675_softSPI_array_Serial_LCD\MAX6675_softSPI_array_Serial_LCD.ino:24:0:
(...)\Documents\Arduino\libraries\MAX6675/MAX6675.h:60:3: note: candidate: MAX6675::MAX6675(uint8_t, uint8_t, uint8_t)
   MAX6675(uint8_t select, uint8_t miso, uint8_t clock);
   ^~~~~~~
(...)\Documents\Arduino\libraries\MAX6675/MAX6675.h:60:3: note:   candidate expects 3 arguments, 0 provided
(...)\Documents\Arduino\libraries\MAX6675/MAX6675.h:58:3: note: candidate: MAX6675::MAX6675(uint8_t, SPIClass*)
   MAX6675(uint8_t select, __SPI_CLASS__ * mySPI);
   ^~~~~~~
(...)\Documents\Arduino\libraries\MAX6675/MAX6675.h:58:3: note:   candidate expects 2 arguments, 0 provided
(...)\Documents\Arduino\libraries\MAX6675/MAX6675.h:54:7: note: candidate: constexpr MAX6675::MAX6675(const MAX6675&)
 class MAX6675
       ^~~~~~~
(...)\Documents\Arduino\libraries\MAX6675/MAX6675.h:54:7: note:   candidate expects 1 argument, 0 provided
(...)\Documents\Arduino\libraries\MAX6675/MAX6675.h:54:7: note: candidate: constexpr MAX6675::MAX6675(MAX6675&&)
(...)\Documents\Arduino\libraries\MAX6675/MAX6675.h:54:7: note:   candidate expects 1 argument, 0 provided
(...)MAX6675_softSPI_array_Serial_LCD\MAX6675_softSPI_array_Serial_LCD.ino: In function 'void ThermoCouplesBegin()':
(...)MAX6675_softSPI_array_Serial_LCD\MAX6675_softSPI_array_Serial_LCD.ino:194:69: error: no matching function for call to 'MAX6675::begin(const int&, int&, const int&)'
     ThermoCouples[i].begin(MAXclockPin, THCselectPins[i], MAXmisoPin);
                                                                     ^
In file included from (...)MAX6675_softSPI_array_Serial_LCD\MAX6675_softSPI_array_Serial_LCD.ino:24:0:
(...)\Documents\Arduino\libraries\MAX6675/MAX6675.h:62:12: note: candidate: void MAX6675::begin()
   void     begin();
            ^~~~~
(...)\Documents\Arduino\libraries\MAX6675/MAX6675.h:62:12: note:   candidate expects 0 arguments, 3 provided

exit status 1

Compilation error: no matching function for call to 'MAX6675::MAX6675()'

Arduino IDE 2.3.2

Why in new versions this happened?

(updated a bit for readability)

**Comment by RobTillaart** [comment]
Thanks for your question,
Will investigate this afternoon and come back.

**Comment by RobTillaart** [comment]
From CHANGELOG.md

```
## [0.3.0] - 2024-01-19
- Fix #9, improve handling SPI dependency.

----

## [0.2.0] - 2023-11-28
- refactor constructor/begin interface - breaking changes.   //  <<<<<<<<<<<<<<<<
  - minimize conditional code. -- create SPI_CLASS macro to solve it.
- update readme.md
- update examples
```

The constructor  begin interface has changed as all my (SPI based) libraries got a wider audience and needed to be able to run on more and different boards. To make that possible the constructor and begin() function were changed to have far less conditional code. This reduces the complexity and improves code quality and maintainability.

That explained, still a breaking change.

**Comment by RobTillaart** [comment]
@rkierzek 
Wrote a small example that compiles (1.8.19), I do not have HW nearby to test.

Please give it a try and let me know if it works

In the example I use a mix of HW SPI and SW SPI, and you have to adjust that to your configuration.
Note that the select pins should be unique.

```cpp
//
//    FILE: MAX6675_array.ino
//  AUTHOR: Rob Tillaart
// PURPOSE: demo array of thermocouples
//     URL: https://github.com/RobTillaart/MAX6675
//     URL: https://github.com/RobTillaart/MAX6675/issues/11

#include "MAX6675.h"

const int ThermoCouplesNum = 4;

MAX6675 ThermoCouples[ThermoCouplesNum] =
{
  MAX6675(3, &SPI),   //  HW SPI
  MAX6675(4, &SPI),   //  HW SPI
  MAX6675(5, 6, 7)    //  SW SPI
};

uint32_t start, stop;

void setup()
{
  Serial.begin(115200);
  Serial.println(__FILE__);
  Serial.print("MAX6675_LIB_VERSION: ");
  Serial.println(MAX6675_LIB_VERSION);
  Serial.println();
  delay(250);

  SPI.begin();

  for (int i = 0; i < ThermoCouplesNum; i++)
  {
    ThermoCouples[i].begin();
    ThermoCouples[i].setSPIspeed(4000000);
  }
}

void loop()
{
  delay(100);

  for (int THCnumber = 0; THCnumber < ThermoCouplesNum; THCnumber++)
  {
    start = micros();
    int status = ThermoCouples[THCnumber].read();
    stop = micros();
    float temp = ThermoCouples[THCnumber].getTemperature();

    Serial.print(millis());
    Serial.print("\tID: ");
    Serial.print(THCnumber);
    Serial.print("\tstatus: ");
    Serial.print(status);
    Serial.print("\ttemp: ");
    Serial.print(temp);
    Serial.print("\tus: ");
    Serial.println(stop - start);

    delay(1000);
  }
}

//  -- END OF FILE --
```

**Comment by RobTillaart** [comment]
Output on an UNO  looks like:
- zero temp due to no sensors
- device 2 is a SW SPI which is much slower than HW SPI

```
MAX6675_LIB_VERSION: 0.3.0

351	ID: 0	status: 0	temp: 0.00	us: 32
1351	ID: 1	status: 0	temp: 0.00	us: 32
2353	ID: 2	status: 0	temp: 0.00	us: 256
3453	ID: 0	status: 0	temp: 0.00	us: 32
4455	ID: 1	status: 0	temp: 0.00	us: 28
5456	ID: 2	status: 0	temp: 0.00	us: 260
6557	ID: 0	status: 0	temp: 0.00	us: 32
7558	ID: 1	status: 0	temp: 0.00	us: 28
8559	ID: 2	status: 0	temp: 0.00	us: 256
````

**Comment by rkierzek** [comment]
@RobTillaart RobTillaart

After RobTillaart code change:
```
MAX6675 ThermoCouples[ThermoCouplesNum] =
{
  MAX6675(4, 3, 2),
  MAX6675(5, 3, 2),
  MAX6675(6, 3, 2)
};
```
sketch compliled and flashed OK. But hardware not working as i want. 
Readings output example:
```
09:45:48.313 -> MAX6675_LIB_VERSION: 0.3.0
09:45:48.313 -> 
09:45:48.673 -> 356	ID: 0	status: 0	temp: 0.00	us: 260
09:45:49.677 -> 1357	ID: 1	status: 0	temp: 0.00	us: 260
09:45:50.671 -> 2359	ID: 2	status: 0	temp: 0.00	us: 260
09:45:51.676 -> 
09:45:51.770 -> 3460	ID: 0	status: 0	temp: 0.00	us: 264
09:45:52.749 -> 4461	ID: 1	status: 0	temp: 0.00	us: 264
09:45:53.751 -> 5463	ID: 2	status: 0	temp: 0.00	us: 260
```

sometimes(rare case about 1of100) status or temp is different than 0.
Same HW setup with old lib version working fine. 
Maybe (long shot) a SPI CPOL/CPHA MODE must be other?

**Comment by RobTillaart** [comment]
Ok, good feedback

Can you test MAX6675_test.ino with a single TC?

**Comment by RobTillaart** [comment]
>  But hardware not working as i want.

This is unclear for me - what do mean with **as I want**?

> sometimes(rare case about 1of100) status or temp is different than 0.

So mostly it just reads zero's 

> Maybe (long shot) a SPI CPOL/CPHA MODE must be other?

Not I'm aware of.
Did it change in the library since your working version? 
If so can you point where it did?

**Comment by RobTillaart** [comment]
@rkierzek 

Possibly swapped pins 2 and 3?

```cpp
MAX6675(uint8_t select, uint8_t miso, uint8_t clock)
```

**Comment by rkierzek** [comment]
Pins MISO and CLK in correct connection.

**Comment by rkierzek** [comment]
Hello, again.

I tested new library with success! All 10 modules together.
Software part is OK. Unfortunately hardware setup is problem.
Connecting additional electrolytic capacitors every couple (3/4/5) MAX6675 modules completely eliminated problems with the stability of readings of some modules. For some reason, unknown to me, I did not immediately notice the problems in the previous library, because not all thermocouples were always analyzed, and not always occur.

Ultimately, the problems I noticed after adapting the code to the new version of the library were hardware related, the fault was all mine.
I am very sorry for the confusion.
The library is very good and useful.
